### ISSUE TITLE
Investigate how the component interacts with similar functionality in browsers

### ISSUE THREAD

**Issue by calvin-lau-sig7** [alphagov/govuk-design-system#3552]
During working group review #3485, suggestion was made to explore how similar functionality in browsers might potentially confuse users. They also suggested suppressing similar functionality if it was possible.

So this work is to assess how the Password input component, particularly the show/hide password button, might interact with similar functionality in browsers.

## What

Make assessment, possibly by desk research, and add findings to the component guidance.
Our findings are likely not to be conclusive as testing across all supported browsers will probably not be practical.

Out of scope:
- browser extensions and plug-ins (as there are too many combinations in the developer marketplace)
- third-party tools such as password managers (as users would've likely had to install or be aware of the tool on their device)

A summary of findings, and our understanding of the issue, should be included in the component guidance #3484 – there is a currently a draft version of this under the 'Research on this component' heading.

## Why

We'd like to get confidence that browser functionality will not cause confusion with our users. 
We'd also like to set the basis for getting more comprehensive feedback about this issue from our community.

## Who needs to work on this
Developer, interaction designer

## Who needs to review this
Wider team

## Done when
- [x]  Investigate how the component interacts with similar functionality in browsers, password managers, screen readers and similar tools
- [x]  Make any needed mitigations into code
- [ ] Add summary of our work and open questions to 'Research on this component' section of the guidance #3355

**Comment by querkmachine** [comment]
Tested in the most recent versions of:

- Apple Safari (macOS, iOS)
- Google Chrome (Windows, macOS, Android)
- Microsoft Edge (Windows, macOS)
- Microsoft Internet Explorer (Windows)
- Mozilla Firefox (Windows, macOS)
- Samsung Internet (Android)

[Testing results grid (internal only).](https://docs.google.com/spreadsheets/d/1WD4kEdquV9h3CIOtnYkW0xHzVhiXTbPEWC8c5TvT5Z8)

Generally speaking, I think we're good. In none of these browsers did the password visibility toggle interfere visually or functionally with the browser's own password management functions (or iCloud Keychain, in the case of Safari).

There are a few browsers where there is the possibility of password visibility not being in sync with what our show/hide button shows, however. In all cases, these are functions of the user agent that—as far as I am aware—cannot be easily overridden by our own code. 

I don't consider any of these to be blocking, but worth noting down just in case.

## Edge (Windows only)

Microsoft Edge has a password visibility toggling function built into it. Although the visual control for this can be hidden by hooking onto the `::-ms-reveal` pseudo-element (and has been), there is an associated keyboard shortcut that cannot be disabled: <kbd>Alt</kbd> + <kbd>F8</kbd>. 

Pressing this shortcut toggles the visibility of a password input, but does not manipulate the DOM or emit JavaScript events, so we cannot syncronise our button to the input's visible state. 

This shortcut only appears to work on Windows.

## Safari (macOS only)

In Safari on macOS, if the user selects "Suggest new password" from the list of presented options, a random password will be generated and displayed within the input in plain text. A gradient overlaps part of the password, obscuring part of it, along with new accompanying text indicating the password strength. 

The input can no longer be directly edited by the user in this situation. Focusing the input instead opens a system dialog with options to regenerate the password.

This does not modify the DOM, though it does alter the shadow DOM. It emits both `input` and `change` events to JavaScript, so we could try and syncronise with this somehow, though I'm not sure if there's any clear way of discerning a value change made by Safari versus one made by a user. 

Clicking the show/hide toggle reverts the shadow DOM changes and whilst retaining the value. 

## Firefox (macOS and Windows)

If the user selects the option to have Firefox generate the password for them, the generated password will be displayed within the input in plain text. It is automatically hidden when the input loses focus, however will become plain text again each subsequent time the input is focused. This persists even if the input's value is modified or the custom show/hide toggle is used.

No DOM changes or JavaScript events are emitted when this toggling happens.

**Comment by selfthinker** [comment]
There are some ways how screen readers can read a password. This list might not be extensive, but this is what I found:

* NVDA has the [Report Password add-on](https://addons.nvda-project.org/addons/reportPasswords.en.html) which announces each character as it is typed (it doesn't read what's in the field)
* NVDA also has the [Speak Password add-on](https://addons.nvda-project.org/addons/speakPasswords.en.html) but that doesn't seem to be compatible with the latest version of NVDA
* I've read about multiple JAWS users who would use Narrator just for the password field, although I found that quite unreliable, but it sometimes works
* Apparently very early on JAWS used to have the functionality to read passwords but it has been removed
* Lots of people open Notepad to write the password in there and then copy & paste it into the password field to make it easier to check what they have typed
* When using a screen reader with a touch screen (like VoiceOver on iOS and TalkBack on Android), characters are always announced when they are typed anyway

@querkmachine started to do some accessibility testing and I finished testing today. The [testing results](https://docs.google.com/spreadsheets/d/1R8co4DEaO1krUs0ZhPFLAjLpS1YIWs1rNix7qmRFoSM/edit?usp=sharing) are in an internal document.

## Screen readers

We tested with JAWS, NVDA, VoiceOver on macOS and iOS and TalkBack. Most screen readers were behaving as expected. The biggest difference was when activating the Show/Hide button, most screen readers only read the ARIA live announcement, and only NVDA read both. As long as either one or the other is read, that should be fine. VoiceOver (VO) on iOS starts to announce the label change but gets interrupted by the ARIA live announcement.

Only VO on macOS causes some issues that **should be further investigated**:

* When using Safari, VO sometimes doesn't read anything at all when the Show/Hide button is clicked. And sometimes it only reads the ARIA live announcement. I couldn't yet figure out why and when the behaviour changes. This doesn't happen in Chrome.
* Sometimes VO doesn't read the label and input type of the password field. It focuses on the first option in the dropdown for Safari's password manager. But sometimes it does read the label and input type fine. Again, I couldn't yet figure out why the behaviour is sometimes different.

## Voice Control

Everything was working as expected when testing with Dragon. I noticed how very useful the show/hide functionality is as it's really difficult to tell what has been inputted all at once when dictating whole words.

## Screen magnifier

I've tested with magnifying the screen to 400% and I had a viewport width of 1280px. With that combination the Show/Hide button is **impossible to notice**. I could see a bit less than half of the width of the password field. Although lots of screen magnification users will have bigger screens and a wider viewport, some will also magnify more than 400%.

**Comment by selfthinker** [comment]
@querkmachine made some changes that I suggested to the ARIA live region. I re-tested how the Show/Hide button behaves in all the screen readers and found that this fixed the first two of the three VO issues.

I have also tested how it would change if we added `aria-pressed` or `role=switch` and `aria-checked`. (See 'testing variants' tab in the same results sheet.) Either seem to make sense. But because this is late in the process, the current solution works well, and we haven't had time to research this change, we decided not to add it for now.
